crates/fuel-core/src/database/balances.rs

```diff
@@ -48,37 +59,101 @@ impl StorageMutate<ContractsAssets> for Database {
         key: &<ContractsAssets as Mappable>::Key,
         value: &<ContractsAssets as Mappable>::Value,
     ) -> Result<Option<<ContractsAssets as Mappable>::OwnedValue>, Self::Error> {
-        Database::insert(self, key.as_ref(), Column::ContractsAssets, value)
-            .map_err(Into::into)
+        let prev = Database::insert(self, key.as_ref(), Column::ContractsAssets, value)
+            .map_err(Into::into);
+
+        // Get latest metadata entry
+        let prev_metadata = self
+            .iter_all::<Vec<u8>, SparseMerkleMetadata>(
+                Column::ContractsAssetsMerkleMetadata,
+                Some(IterDirection::Reverse),
+            )
+            .next()
+            .transpose()?
+            .map(|(_, metadata)| metadata)
+            .unwrap_or_default();
+
+        let root = prev_metadata.root;
+        let mut tree: MerkleTree<ContractsAssetsMerkleData, _> = {
+            let storage = self.borrow_mut();
+            if root == [0; 32] {
+                // The tree is empty
+                MerkleTree::new(storage)
+            } else {
+                // Load the tree saved in metadata
+                MerkleTree::load(storage, &root)
+                    .map_err(|err| StorageError::Other(err.into()))?
+            }
+        };
+
+        // Update the key-value dataset. The key is the contract id and the
```

Comment:
Hmm, is that true? The value is not 32 bytes, it is `Word` - 8 bytes

Reply:
Yep, thanks for the catch!